File: src/lib/markdown.js
/**
 * Lightweight YAML frontmatter parser/serializer.
 * No external dependencies — regex-based.
 */

const FRONTMATTER_RE = /^---\r?\n([\s\S]*?)\r?\n---\r?\n?([\s\S]*)$/;

export function parseFrontmatter(content) {
  const match = content.match(FRONTMATTER_RE);
  if (!match) return { data: {}, body: content };

  const raw = match[1];
  const body = match[2];
  const data = {};

  for (const line of raw.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;

    const colonIdx = trimmed.indexOf(':');
    if (colonIdx === -1) continue;

    const key = trimmed.slice(0, colonIdx).trim();
    const rawVal = trimmed.slice(colonIdx + 1).trim();
    data[key] = parseValue(rawVal);
  }

  return { data, body };
}

function parseValue(raw) {
  if (raw === '' || raw === 'null') return null;
  if (raw === 'true') return true;
  if (raw === 'false') return false;
  if (/^\d+$/.test(raw)) return parseInt(raw, 10);
  if (/^\d+\.\d+$/.test(raw)) return parseFloat(raw);

  if (raw.startsWith('[') && raw.endsWith(']')) {
    return raw
      .slice(1, -1)
      .split(',')
      .map(s => s.trim())
      .filter(Boolean)
      .map(s => parseValue(s));
  }

  if ((raw.startsWith('"') && raw.endsWith('"')) ||
      (raw.startsWith("'") && raw.endsWith("'"))) {
    return raw.slice(1, -1);
  }

  return raw;
}

export function serializeFrontmatter(data, body) {
  const lines = ['---'];
  for (const [key, val] of Object.entries(data)) {
    lines.push(`${key}: ${serializeValue(val)}`);
  }
  lines.push('---');
  if (body) lines.push('', body);
  return lines.join('\n');
}

function serializeValue(val) {
  if (val === null || val === undefined) return 'null';
  if (typeof val === 'boolean') return String(val);
  if (typeof val === 'number') return String(val);
  if (Array.isArray(val)) {
    if (val.length === 0) return 'null';
    return val.join(', ');
  }
  return String(val);
}

File: projctl/lib/markdown.py
from __future__ import annotations
import re
from typing import Any

_FRONTMATTER_RE = re.compile(r"^---\r?\n([\s\S]*?)\r?\n---\r?\n?([\s\S]*)$")


def parse_frontmatter(content: str) -> tuple[dict, str]:
    match = _FRONTMATTER_RE.match(content)
    if not match:
        return {}, content

    raw, body = match.group(1), match.group(2)
    data: dict = {}

    for line in raw.splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        idx = line.find(":")
        if idx == -1:
            continue
        key = line[:idx].strip()
        raw_val = line[idx + 1:].strip()
        data[key] = _parse_value(raw_val)

    return data, body


def _parse_value(raw: str) -> Any:
    if raw in ("", "null"):
        return None
    if raw == "true":
        return True
    if raw == "false":
        return False
    if re.fullmatch(r"\d+", raw):
        return int(raw)
    if re.fullmatch(r"\d+\.\d+", raw):
        return float(raw)
    if raw.startswith("[") and raw.endswith("]"):
        inner = raw[1:-1]
        return [_parse_value(s.strip()) for s in inner.split(",") if s.strip()]
    if (raw.startswith('"') and raw.endswith('"')) or \
       (raw.startswith("'") and raw.endswith("'")):
        return raw[1:-1]
    return raw


def serialize_frontmatter(data: dict, body: str = "") -> str:
    lines = ["---"]
    for key, val in data.items():
        lines.append(f"{key}: {_serialize_value(val)}")
    lines.append("---")
    if body:
        lines.append("")
        lines.append(body)
    return "\n".join(lines)


def _serialize_value(val: Any) -> str:
    if val is None:
        return "null"
    if isinstance(val, bool):
        return str(val).lower()
    if isinstance(val, (int, float)):
        return str(val)
    if isinstance(val, list):
        if not val:
            return "null"
        return ", ".join(str(v) for v in val)
    return str(val)
